### PR TITLE
[engine] call: ResultNeedUpgradeVerification

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances #17082

Engine calls `ResultNeedUpgradeVerification` when both:
- (a) the upgrading feature flag is enabled
- (b) the source/destination template types differ (i.e. an upgrade or downgrade has occurred)

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -1200,23 +1200,6 @@ private[lf] object SBuiltin {
     }
   }
 
-  final case class SBGuardMatchTemplateId(
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1203]: (unrelated dead code)

**Review by remyhaemmerle-da — APPROVED** [review]
